src/main/java/core/basesyntax/db/FruitDaoImpl.java

```diff
@@ -0,0 +1,47 @@
+package core.basesyntax.db;
+
+import core.basesyntax.exception.InvalidFruitDataException;
+import java.util.Map;
+
+public class FruitDaoImpl implements FruitDao {
+    private static final int ZERO = 0;
+
+    @Override
+    public Integer setFruitBalance(String fruitName, int quantity) {
+        String fruitKey = fruitName.toLowerCase();
+
+        return FruitDB.getFruitDataBase().merge(fruitKey, quantity, (oldV, newV) -> newV);
+    }
+
+    @Override
+    public Integer getFruitBalance(String fruitName) {
+        String fruitKey = fruitName.toLowerCase();
+
+        return FruitDB.getFruitDataBase().getOrDefault(fruitKey, ZERO);
+    }
+
+    @Override
+    public Integer addFruits(String fruitName, int quantity) {
+        String fruitKey = fruitName.toLowerCase();
+
+        return FruitDB.getFruitDataBase().merge(fruitKey, quantity, Integer::sum);
+    }
+
+    @Override
+    public Integer subtractFruits(String fruitName, int quantity) {
+        String fruitKey = fruitName.toLowerCase();
+
+        if (!FruitDB.getFruitDataBase().containsKey(fruitKey)) {
+            throw new InvalidFruitDataException("There is no such fruit in the database: "
+                    + fruitName);
+        } else {
+            return FruitDB.getFruitDataBase()
+                    .merge(fruitKey, quantity, (oldV, newV) -> oldV - newV);
+        }
```

Comment:
would be perfect if you set merging at the moment of creating of HashMap. you can pass merge algorithm as a constructor parameter, and there will be no need to call it each time after that